### DISCUSSION TITLE
lightsleep, deepsleep & wifi

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12092] [Using MicroPython]
When using the esp32 to send remote sensor data every 20 mins or so I generally use deepsleep between updates. I originally did this to reduce battery drain but it also means I don't have to think too much about 802.11 complexity since the program connects to the wifi after each deepsleep. The problem is 3s of the 7s awake time each cycle is the time taken to connect to wifi. 

I'd like to try lightsleep to see if not having to reconnect to the wifi each cycle could be worth a look. I've tried reading up on listen intervals & the requirement for APs to buffer frames for dozing stations but it's heavy going & none of it is esp32 specific. Does the esp32 WLAN keep itself connected in lightsleep? If so how often is it likely to wake up to listen for beacons or do whatever it does to stay connected?

**Comment by peterhinch** [comment]
See [this doc](https://github.com/glenn20/upy-esp32-experiments) by @glenn20 on reducing energy used on boot.

**Comment by sosi-deadeye** [comment]
> Does the esp32 WLAN keep itself connected in lightsleep? If so how often is it likely to wake up to listen for beacons or do whatever it does to stay connected?

Sometimes yes and sometimes no. I tested it with an ESP32 with 60 seconds. I was able to make an HTTP GET request after the `lightsleep` without reconnecting. After some tries, I got an OSError -> no connection

So even after a `lightsleep` you may have to reconnect `WLAN`.

**Reply by karfas** [reply]
Please check out the ESP-IDF docs regarding sleep modes and WIFI.

They explicitly require to switch off any WLAN and Bluetooth functions **before** going to **any** sleep mode, so I think this is not supported at all.

**Reply by glenn20** [reply]
Yes - there is some confusing language used about "sleep" in the esp-idf docs. `lightsleep` and `deepsleep` refer to the sleep states of the device (CPU). As noted by @karfas you are **supposed** to put the radio to sleep whenever entering a sleep mode, but you can get away with not doing so in some/many cases (as experienced by @sosi-deadeye) - and you mileage might depend on your AP and how it is configured.

But sleep is also used to refer to putting the radio to sleep (sometimes called `modem-sleep`) while connected to a wifi access point. On the esp32, the wifi has three power saving modes:
- `WLAN.config(ps=WLAN.PM_PERFORMANCE)`: which set the default power saving mode 
   - Turn on the wifi radio every DTIM interval to receive data from the wireless access point.
   - The DTIM interval is set by your wifi access point (typically set to 1 Beacon interval (typically 100ms), but can be increased to 2, 3, 4, ... beacon intervals.
- `WLAN.config(ps=WLAN.PM_POWERSAVE)`:
   - Turn on the wifi radio every 3 DTIM interval to receive data from the wireless access point (you can tune this in the IDF, but micropython does not support that - and it is a case of diminishing returns to increase it further).
 - `WLAN.config(ps=WLAN.PM_NONE)`: disable the power saving altogether and just leave the radio on whenever either interface is `active(True)`.
 
When you connect to an access point, PM_PERFORMANCE mode is activated automatically by the ESP32.

Note, that the radio is only turned off after the wifi has been idle for some period of time (eg. a few seconds) and the radio is turned on immediately if you send a packet.

Typically the powersave mode might only provide a small additional power saving over the performance mode. In these modes, the ESP32 is fully powered on, it is just the radio being switched on and off.

You **can** change the Beacon interval and the DTIM interval on your wifi access point. If you set the DTIM much larger, it can save power on the DTIM-enabled devices attached to your network, but some applications and other devices may no longer work well.

The second graph on [this page](https://github.com/glenn20/upy-esp32-experiments/blob/main/ESPNowPowerManagement/README.md) (referenced by @peterhinch) shows a power profile very similar to the PM_PERFORMANCE mode (that graph is for a hand-tuned ESPNOW power setting, but is almost identical to the profile when connected to my wifi AP (Beacon=100ms, DTIM=1). You can clearly see the radio turning off and how much power is saved.

**Reply by kjm1102** [reply]
Glenn, when you say 'a few seconds' for wifi idle you mean the esp32 radio is turned off?  Wouldn't that mean it could miss data sent to it by the AP during that time?

**Reply by glenn20** [reply]
There is a protocol for this. Wifi access points buffer up data for clients and send them when they are ready to receive them  (the DTIM interval). This lets devices like phones reduce battery consumption when connected to wifi at the cost of some latency in communications. The ESP32 automatically turns on this mode when the STA_IF interface connects to an access point. The DTIM interval is typically 100ms or so. Increasing the wake interval further will increase latency further and mean that your device may miss wifi broadcast messages.

**Comment by bidrohini** [comment]
I usually see these write-ups to see which parts of the ESP32 are active during which low mode:

https://lastminuteengineers.com/esp32-sleep-modes-power-consumption/
https://www.theengineeringprojects.com/2021/12/esp32-low-power-modes.html

It looks like only the ULP co-processor and the RTC are active during the light sleep mode, nothing else......

**Reply by kjm1102** [reply]
My devkit wroom board with v1.20 upython is happy to stay connected to wifi through 20 minute light sleeps no prob.

**Reply by bidrohini** [reply]
Oh! That was something unknown to me. Thanks.

**Reply by glenn20** [reply]
During `machine.lightsleep()` the wifi radio is turned off, so I guess the behaviour after you wake might depend on whether the AP considers you to be still connected. 

I just ran some tests: 

- If I `machine.lightsleep(60_000); sta.isconnected()` (60 second lightsleep) **without** calling `sta.active(False)`, my ESP32 (DFRobot Beetle ESP32 v1.0 - also a wroom board) is disconnected after waking from the AP about 10-50% of the time.
  - On some of those occasions I am unable to reconnect to the AP until I restart the wifi (`sta.active(False); sta.active(True)`).

This behaviour **may** depend on how your Access Point behaves.

**Note:**: The [espressif docs](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/sleep_modes.html?highlight=deepsleep#wi-fi-bluetooth-and-sleep-modes) say you must turn off the wifi before entering lightsleep, but clearly you can get away with it some or most of the time, but perhaps not always.

**Comment by robert-hh** [comment]
I do not know why you look at secondary sources when the good primary source of information is available: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/sleep_modes.html
In all questions about ESP32 it's best to look at the espressif pages first. Their documentation is good.
Besides that, a firmware like MicroPython may configure the modes differently.

**Reply by kjm1102** [reply]
Let's not quibble about sources Rob. Maybe apply your expertise to https://github.com/orgs/micropython/discussions/12106 instead?

**Comment by bidrohini** [comment]
> I do not know why you look at secondary sources when the good primary source of information is available: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/sleep_modes.html In all questions about ESP32 it's best to look at the espressif pages first. Their documentation is good. Besides that, a firmware like MicroPython may configure the modes differently.

Thank you. I am saving it.
